contracts/casino/src/casino.cpp

```diff
@@ -57,39 +68,76 @@ void casino::on_transfer(name game_account, name casino_account, eosio::asset qu
+    const auto game_id = get_game_id(game_account);
     verify_game(game_id);
     transfer(player_account, quantity);
     sub_balance(game_id, quantity);
 }
 
 void casino::claim_profit(name game_account) {
-    const auto ct = eosio::current_time_point();
+    const auto ct = current_time_point();
     const auto game_id = get_game_id(game_account);
     verify_game(game_id);
     check(ct - get_last_claim_time(game_id) > microseconds(useconds_per_month), "already claimed within past month");
-
-    const auto game_row = platform::read::get_game(platform_contract, game_id);
-    const auto balance = get_balance(game_id);
-    const auto profit_margin = game_row.profit_margin;
-    const auto beneficiary = game_row.beneficiary;
-    const auto to_transfer = balance * profit_margin / 100;
+    const auto beneficiary = platform::read::get_game(platform_contract, game_id).beneficiary;
+    const auto to_transfer = get_game_profits(game_id);
     transfer(beneficiary, to_transfer);
-
     sub_balance(game_id, to_transfer);
     update_last_claim_time(game_id);
 }
 
-
 uint64_t casino::get_game_id(name game_account) {
     // get game throws if there's no game in the table
     return platform::read::get_game(platform_contract, game_account).id;
 }
 
-
 void casino::verify_game(uint64_t game_id) {
     check(platform::read::is_active_game(platform_contract, game_id), "the game was not verified by the platform");
     check(is_active_game(game_id), "the game is not run by the casino");
 }
 
+void casino::verify_account(name game_account) {
+    verify_game(get_game_id(game_account));
+}
+
+void casino::withdraw(name beneficiary_account, asset quantity) {
+    require_auth(get_owner());
+    const auto ct = current_time_point();
+    const auto account_balance = token::get_balance(_self, core_symbol);
+
+    if (account_balance > gstate.game_active_sessions_sum + gstate.game_profits_sum) {
+        const asset max_transfer = account_balance - gstate.game_active_sessions_sum - gstate.game_profits_sum;
+        check(quantity <= max_transfer, "quantity exceededs max transfer amount");
+        transfer(beneficiary_account, quantity);
+    } else {
+        check(account_balance > gstate.game_profits_sum, "developer profits exceed account balance");
+        const asset max_transfer = account_balance / 10;
+        check(quantity <= max_transfer, "quantity exceededs max transfer amount");
+        check(ct - gstate.last_withdraw_time > microseconds(useconds_per_week), "already claimed within past week");
+        transfer(beneficiary_account, quantity);
+        gstate.last_withdraw_time = ct;
+    }
+}
+
+void casino::session_update(name game_account, asset max_win_delta) {
+    require_auth(game_account);
+    verify_account(game_account);
+    session_update(max_win_delta);
+}
+
+void casino::session_close(name game_account, asset quantity) {
+    require_auth(game_account);
+    verify_account(game_account);
```

Comment:
Paused game should be able to close session(e.g in case when game become paused after session start)

Reply:
on the other hand this means every account can invoke this action, casino itself for example

Reply:
why ?
you still leave all other checks including game listing check and sender check

Reply:
ok

Reply:
fixed